third_party/libunwindstack/PeCoffEpilog.cpp

```diff
@@ -0,0 +1,306 @@
+    last_error_.address = registers->sp();
+    return false;
+  }
+  registers->set_sp(registers->sp() + sizeof(uint64_t));
+  registers->set_pc(return_address);
+
+  return true;
+}
+
+bool PeCoffEpilog::ValidateJumpInstruction() {
+  // It's not entirely clear how to distinguish between regular 'jmp' instructions and 'jmp'
+  // instructions that are at the end of an epilog (e.g. due to tail call optimization).
+  // There are some restrictions which 'jmp' instructions are allowed in epilogs, but this
+  // doesn't solve the problem of distinguishing entirely. This means that we may identify all
+  // 'jmp' instructions that satisfy this restriction as an epilog consisting of a single
+  // instruction.
+  // TODO: We may need to look at the unwind codes of this function to see if the epilog should
+  // be non-trivial (and not just consist of a single 'jmp' instruction).
+
+  // Only 'jmp' instructions with memory references are allowed in the epilog:
+  // https://docs.microsoft.com/en-us/cpp/build/prolog-and-epilog?view=msvc-170#epilog-code
+  CHECK(capstone_instruction_->detail->x86.op_count >= 1);
+  cs_x86_op operand0 = capstone_instruction_->detail->x86.operands[0];
+  if (operand0.type != X86_OP_MEM) {
+    return false;
+  }
+
+  // Only instructions with mod = 0b00 are allowed according to
+  // https://docs.microsoft.com/en-us/cpp/build/prolog-and-epilog?view=msvc-170#epilog-code
+  // (The modrm byte consist of fields mod, reg, and rm, where mod is 2 bits, reg is 3 bits, and
+  // rm is 3 bits.)
+  if (capstone_instruction_->detail->x86.modrm & 0xc0) {
+    return false;
+  }
+
+  return true;
+}
+
+bool PeCoffEpilog::HandleJumpInstruction(Memory* process_memory, RegsImpl<uint64_t>* registers) {
+  // Seeing a 'jmp' at the end of the epilog means we are jumping into some other function  that
+  // will carry out prolog and at the end epilog instructions, setting up and unwinding a
+  // callframe. We do not have to simulate all these steps by the function we are jumping to,
+  // the return address leading back to the function that called the current function is already
+  // on the top of the stack and we can just directly virtually return here. Hence, the rest of
+  // the code to handle this case is exactly the same as in the 'ret' case.
+  uint64_t return_address;
+  if (!process_memory->Read64(registers->sp(), &return_address)) {
+    last_error_.code = ERROR_MEMORY_INVALID;
+    last_error_.address = registers->sp();
+    return false;
+  }
+  registers->set_sp(registers->sp() + sizeof(uint64_t));
+  registers->set_pc(return_address);
+
+  return true;
+}
+
+// Disassembles the machine code passed in, and scans through the instructions one-by-one to detect
+// if the machine code is an epilog according to the specification. For x86_64 on on Windows,
+// epilogs must follow a specific pattern as described on:
+// https://docs.microsoft.com/en-us/cpp/build/prolog-and-epilog?view=msvc-170
```

Comment:
I guess what is now missing is a rough idea about what is done in the "handle" parts of the function.

Reply:
Done, added a sentence, though I want to be careful not to add too many comments that just describe what the code is doing.

Reply:
Thanks. And yes agree with that. But it felt like it was missing when reading through the comment.